Revendored Swarmkit

RenaudWasTaken opened the pull request:
includes https://github.com/docker/swarmkit/pull/2419, which:

fixes https://github.com/moby/moby/issues/34825
fixes https://github.com/moby/moby/issues/34996
fixes https://github.com/moby/moby/issues/35046
fixes https://github.com/moby/moby/issues/35395


Hello!

**- What I did**
- Revendored swarmkit to the latest commit to include the recent GenericResource change
- Updated the Generic resource CLI to match it's new CLI format
- Fixed the DockerSwarmSuite.TestSwarmVolumePlugin as the error message is now surfaced in a new field

**- Description for the changelog**
Revendored Swarmkit

@anshulpundir can you confirm that the revendoring does not require me to change moby/moby ?
Ping @thaJeztah

**- Bumps**

https://github.com/docker/swarmkit/compare/28f91d87bd3f75fd039dbb9be49bfd2381019261...de950a7ed842c7b7e47e9451cde9bf8f96031894

- Delete node attachments when node is removed https://github.com/docker/swarmkit/pull/2409
- add pid limits https://github.com/docker/swarmkit/pull/2415 (related to https://github.com/moby/moby/issues/28618)
- Provide custom gRPC dialer to override default proxy dialer https://github.com/docker/swarmkit/pull/2419
- Generic resource cli https://github.com/docker/swarmkit/pull/2347


**- A picture of a cute animal (not mandatory but encouraged)**
![jack-russell-2-1024x683](https://user-images.githubusercontent.com/2519571/32135657-ef17fae2-bc02-11e7-84db-77b9fa1ea5f0.jpg)

RenaudWasTaken committed:
Revendored swarmkit

Signed-off-by: Renaud Gaubert <renaud.gaubert@gmail.com>

yongtang commented:
Looks like the following failure happens on all platform:
```
14:55:19 FAIL: docker_cli_swarm_unix_test.go:15: DockerSwarmSuite.TestSwarmVolumePlugin
14:55:19 
14:55:19 [d539f05da0b31] waiting for daemon to start
14:55:19 [d539f05da0b31] daemon started
14:55:19 
14:55:19 docker_cli_swarm_unix_test.go:22:
14:55:19     // Make sure task stays pending before plugin is available
14:55:19     waitAndAssert(c, defaultReconciliationTimeout, d.CheckServiceTasksInState("top", swarm.TaskStatePending, "missing plugin on 1 node"), checker.Equals, 1)
14:55:19 docker_utils_test.go:465:
14:55:19     c.Assert(v, checker, args...)
14:55:19 ... obtained int = 0
14:55:19 ... expected int = 1
14:55:19 
14:55:19 [d539f05da0b31] exiting daemon
14:55:20 
```

RenaudWasTaken committed:
Updated GenericResource CLI

Signed-off-by: Renaud Gaubert <renaud.gaubert@gmail.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "swarmkit-revendored" git@github.com:RenaudWasTaken/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354008608
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

RenaudWasTaken commented:
After some work I found that the root cause of this bug was introduced by https://github.com/docker/swarmkit/pull/2287

The error message is now surfaced in the Error field. I've added a commit to fix this,

thaJeztah commented:
Thanks! There's another PR as well that bumps swarm kit, so one of both will have to rebase probably. I created a quick overview of changes included in this PR (currently);

https://github.com/docker/swarmkit/compare/872861d2ae46958af7ead1d5fffb092c73afbaf0...de950a7ed842c7b7e47e9451cde9bf8f96031894

From that list; included in this bump are:

- api: Use TaskStatus Err field for non-terminal errors https://github.com/docker/swarmkit/pull/2287
- Support the init flag for container spec https://github.com/docker/swarmkit/pull/2386
- [Windows] Support for isolation mode on container spec https://github.com/docker/swarmkit/pull/2342 (related to https://github.com/moby/moby/pull/34424)
- Increase gRPC request timeout to 20 seconds for sending snapshots https://github.com/docker/swarmkit/pull/2391
- Delete node attachments when node is removed https://github.com/docker/swarmkit/pull/2409
- add pid limits https://github.com/docker/swarmkit/pull/2415 (related to https://github.com/moby/moby/issues/28618)
- Generic resource cli https://github.com/docker/swarmkit/pull/2347

thaJeztah commented on daemon/config/config.go:
```diff
@@ -171,7 +171,8 @@ type CommonConfig struct {
 	// Exposed node Generic Resources
-	NodeGenericResources string `json:"node-generic-resources,omitempty"`
+	// e.g: ["orange=red", "orange=green", "orange=blue", "apple=3"]
+	NodeGenericResources []string `json:"node-generic-resources,omitempty"`
```
Do we have a test to test loading this configuration from a `daemon.json` configuration file?

RenaudWasTaken replied:
Just added one :)

RenaudWasTaken committed:
Added NodeGenericResource config tests

Signed-off-by: Renaud Gaubert <renaud.gaubert@gmail.com>

anshulpundir commented:
Should we also include https://github.com/docker/swarmkit/commit/312be5983e2d2f01f0cbcc99d027b28433256705 ?@thaJeztah

anshulpundir reviewed (commented):
Should we also include https://github.com/docker/swarmkit/commit/312be5983e2d2f01f0cbcc99d027b28433256705? @thaJeztah

RenaudWasTaken commented:
Thanks @thaJeztah for the list!
Including it and https://github.com/docker/swarmkit/pull/2419 in the first post.

thaJeztah commented:
@anshulpundir ah! yes, missed that one in my overview; I updated the list, and updated the list at the top as well

@RenaudWasTaken https://github.com/moby/moby/pull/34424 was merged, so this PR now needs a rebase 😅; probably just amending the first commit ("Revendored swarmkit "), and running `vndr` again should do

RenaudWasTaken commented:
Looks like it also fixed the integration test: https://github.com/moby/moby/pull/34424/files#diff-eb3f7d0d8dfa41106bb74d7e5dfe43ecR203

I've removed that commit.
Thanks for updating my first post!

RenaudWasTaken commented:
@thaJeztah is this PR good to go ?
Hoping to revendor the CLI after this so we can get support of GPUs in swarm!

thaJeztah commented on opts/opts.go:
```diff
@@ -263,6 +263,16 @@ func ValidateLabel(val string) (string, error) {
+// i.e 'GPU=UID1' is valid however 'GPU:UID1' or 'UID1' isn't
+func ValidateSingleGenericResource(val string) (string, error) {
+	if strings.Count(val, "=") < 1 {
+		return "", fmt.Errorf("bad attribute format `%s` expected argument of the for `name=value`", val)
```
This message looks incomplete ("argument of the for .. ")?

Looking at other messages, I think we use `invalid ...` for other similar messages (e.g. https://github.com/moby/moby/blob/d91c5f42eb37c6f88cec4021c10c0a1ded1785c3/daemon/cluster/convert/container.go#L296)

Perhaps something like:

```
invalid node-generic-resource format: expected 'name=value':
```

@mistyhacks perhaps you have a better suggestion?

mistyhacks replied:
I think it's ok, if you also include a `got` portion to the error message.  What is a "node-generic-resource"?

RenaudWasTaken replied:
@mistyhacks the full design document [is here](https://github.com/docker/swarmkit/blob/master/design/generic_resources.md).
The basic idea is that it's a way for a node to advertise user-defined resources

thaJeztah reviewed (changes requested):
Thanks! Just some minor nits, but looks good otherwise

thaJeztah reviewed (approved):
LGTM, thanks!

vdemeester reviewed (approved):
LGTM 🐮

thaJeztah commented:
Thanks @RenaudWasTaken - I think this needs an update to the documentation (as the format for the `dockerd` options changed); unfortunately those are not in this repository, but could you open a pull request for the `dockerd` command-line reference, and the man-page in the https://github.com/docker/cli repository?

- https://github.com/docker/cli/blob/master/docs/reference/commandline/dockerd.md
- https://github.com/docker/cli/blob/master/man/dockerd.8.md

RenaudWasTaken commented:
@thaJeztah will be doing this as part of docker/cli/pull/429

thaJeztah commented:
Perfect, thanks @RenaudWasTaken

RenaudWasTaken commented:
Looks like I'll have to revendor ~100 commits of docker/docker though :(

thaJeztah commented:
If you prefer, I can do a separate bump for all changes before yours

RenaudWasTaken commented:
@thaJeztah It would make things a lot easier :) Thanks!

On the other hand it does look like I'll still have to revendor docker  because of the way [parsing is done](https://github.com/moby/moby/blob/f4d4f5863156b82ef146b6ff1e845f8dcf019f12/daemon/config/opts.go#L10).

Basically parsing in the daemon is through swarmkit/api/genericresource.Parse and then docker/daemon/cluster/convert.GenericResourcesFromGRPC

Meaning that if I want to do the same in the CLI I'll need to add a lot of new dependencies. Because that'll mean vendoring docker/daemon/cluster/convert which has a lot of dependencies.

Should I copy swarmkit/api/genericresource.Parse in docker/cli or add the 10+ dependencies to the vendor.conf ?

dnephin commented:
The dependencies of `swarmkit/api/genericresource/` are fine, it's just `swarmkit/api` which is already a dependency of `docker/cli`.

I would duplicate `convert.GenericResourcesFromGRPC()`. There's not much logic there, and `daemon/cluster/convert` is not an appropriate dependency for `docker/cli`.
